src/mp3.c

```diff
@@ -0,0 +1,468 @@
+** Public function.
+*/
+
+int
+mp3_open (SF_PRIVATE *psf)
+{	int	subformat, error = 0 ;
+	MP3_PRIVATE* mdata ;
+	if ((mdata = calloc (1, sizeof (MP3_PRIVATE))) == NULL)
+		return SFE_MALLOC_FAILED ;
+	psf->container_data = mdata ;
+
+	if (psf->file.mode == SFM_RDWR)
+		return SFE_BAD_MODE_RW ;
+
+
+	if (psf->file.mode == SFM_WRITE)
+	{
+
+		return mp3_open_lame (psf) ;
+	}
+
+	if (psf->file.mode == SFM_RDWR)
+		return SFE_UNIMPLEMENTED ;
+
+
+	if (psf->file.mode == SFM_READ)
+		return SFE_UNIMPLEMENTED ;
+
+	return error ;
+} /* mp3_open */
+
+/*------------------------------------------------------------------------------
+*/
+
+static int
+mp3_open_lame (SF_PRIVATE *psf)
+{	MP3_PRIVATE *p = (MP3_PRIVATE*) psf->container_data ;
+	lame_global_flags *gfp = p->gfp = lame_init () ;
+	int error = 0, format, mode ;
+
+	format = SF_CONTAINER (psf->sf.format) ;
+	if (format != SF_FORMAT_MP3)
+		return SFE_BAD_OPEN_FORMAT ;
+	lame_set_num_channels (gfp, psf->sf.channels) ;
+	lame_set_in_samplerate (gfp, psf->sf.samplerate) ;
+	lame_set_brate (gfp, 256) ;	/* FIXME to parameter */
+	mode = (psf->sf.format-SF_FORMAT_MP3_STEREO) & SF_FORMAT_SUBMASK ;
+	lame_set_mode (gfp, mode) ;
+	p->quality = 2 ;
+	//lame_set_quality (gfp, 2) ;	/* 2=high	5 = medium  7=low */
+	if ((p->mp3buffer = calloc (1, MP3BUFFER_SIZE)) == NULL)
+		return SFE_MALLOC_FAILED ;
+	psf->command = mp3_command ;
+	psf->datalength = 0 ;	/* What is this?? */
+	psf->dataoffset = 0 ;
+	if (mode == 3)		/* MONO may overwrite data */
+	{	psf->write_short = mp3_lame_write_short_M ;
+		psf->write_int = mp3_lame_write_int_M ;
+		psf->write_float = mp3_lame_write_float_M ;
+		psf->write_double = mp3_lame_write_double_M ;
+		if ((p->mp3data = calloc (2, MP3DATA_SIZE*sizeof (double))) == NULL)
+			return SFE_MALLOC_FAILED ;
+		}
+	else
+	{	psf->write_short = mp3_lame_write_short ;
+		psf->write_int = mp3_lame_write_int ;
+		psf->write_float = mp3_lame_write_float ;
+		psf->write_double = mp3_lame_write_double ;
+		}
+	psf->container_close = mp3_close_lame ;
+	psf->command = mp3_command ;
+	return error ;
+}
+
+static int
+mp3_close_lame (SF_PRIVATE *psf)
+{	MP3_PRIVATE *p = psf->container_data ;
+	lame_global_flags *gfp = p->gfp ;
+	int bytes = lame_encode_flush (gfp, p->mp3buffer, MP3BUFFER_SIZE) ;
+	if (bytes > 0) psf_fwrite (p->mp3buffer, 1, bytes, psf) ;
+	//lame_mp3_tags_fid (gfp, p->fout) ;
+	lame_close (gfp) ;
+
+	free (p->mp3buffer) ;
+	p->gfp = NULL ;
+	return 0 ;
+}
+
+/*------------------------------------------------------------------------------
+*/
+
+static int
+mp3_lame_parameters (lame_global_flags *gfp, MP3_PRIVATE *p)
+{	int error ;
+	lame_set_quality (gfp, p->quality) ;
+	if ((error = lame_init_params (gfp)) < 0)
+		return SFE_BAD_OPEN_FORMAT ;
+	p->initialised = 1 ;
+	return 1 ;
+}
+
+/* **** Note that if mode is MONO (3) then left and right channels are averaged and the data is changed.  Hence the _M functions **** */
```

Comment:
This stereo mix down to mono functionality should probably be dropped.

Reply:
So you suggest mono is signalled only by sf.channels?  That still leaves the two types of stereo to resolve.  I do need to ensure sf.channels is 1 or 2 only.

Reply:
> So you suggest mono is signalled only by sf.channels? 

Yes.

> I do need to ensure sf.channels is 1 or 2 only.

A number of formats already require things like that. See `sf_format_check` in `src/sndfile.c`.